### PR TITLE
[ETL-423] Allow for self referencing schemas

### PR DESCRIPTION
**Pull request by rxu17**
**Purpose:** This PR allows for s3 to json to have the ability to take in new assessments (and old assessments with new revisions) with self referencing schemas and pull their JSON schemas directly instead of the previous archive map method. See [Self-Referencing Data](https://sagebionetworks.jira.com/wiki/spaces/BD/pages/2883715106/Self-Referencing+Data) for more information on the design and context. 

New function: 
- `get_self_ref_schema_list`

Other relevant changes:
- `get_json_schema`
- Anywhere `get_self_ref_schema_list` is used

New tests added to s3 to json (ignore all `black` linting changes)

### CODE REVIEW

**Comment by philerooski** [comment]
I added an additional logging statement and a workaround for a jsonschema bug. The bug was triggered by the format of the new assessment's `$id` property (e.g., `schemas/v0/3DRotation.json`). `jsonschema` tries to parse the URI as a URL. The issue is explained pretty well in this bug report [here](https://github.com/python-jsonschema/jsonschema/issues/1012):

> However, jsonschema fails here. It looks like two things are needed for the error to happen:
>
> A URI in the "$id" field where jsonschema doesn't know the scheme (or a string that looks like a URI).
Nested "$ref" statements.
>
> According to the spec:
>
>> The [$id](https://json-schema.org/draft/2020-12/json-schema-core.html#section-8.2.1) keyword defines a URI for the schema, and the base URI that other URI references within the schema are resolved against.
>
> jsonschema validates the file just fine if the "$id" doesn't contain a colon. I don't even understand why it tries to resolve the "$id".

In our case, rather than `$id` containing a confusing colon, it contains confusing forward slashes.

```
RefResolutionError: unknown url type: 'schemas/v0/schemas/v0/3DRotation.json'
```

To workaround, I check for this specific `$id` format before instantiating the validator. Apparently jsonschema has made some [big improvements](https://github.com/python-jsonschema/jsonschema/pull/1049) to how they resolve `$ref`s lately. We'll want to explore updating to the latest version of jsonschema in the future.

One additional change I made was to the `get_self_ref_schema_list` function. Technically, we should be validating `metadata.json` against its schema so that we know that it conforms to the format we expect, but adding an extra conditional achieves the same effect – so that any record, whether its `metadata.json` conforms to a specific format or even whether it has a `metadata.json` file in the first place, can be run through the function without raising an error.

**Review by philerooski — APPROVED** [review]

**Review by thomasyu888 — COMMENTED** [review]
🔥 awesome work here! I'm going to leave this to Phil's approval, but just some comments.

File: src/glue/jobs/s3_to_json_s3.py [hunk 1]
```diff
@@ -617,39 +717,40 @@ def main():
...
-            ["WORKFLOW_NAME",
-             "WORKFLOW_RUN_ID",
-             "ssm-parameter-name",
-             "dataset-mapping",
-             "schema-mapping",
-             "archive-map-version",
-             "invalid-sqs"])
+        sys.argv,
+        [
+            "WORKFLOW_NAME",
```
[the remark this answers is not in the document]
Comment (philerooski) [reply src/glue/jobs/s3_to_json_s3.py:726]: I'm not sure we get a choice in this. This is how Glue passes job arguments.

[the remark this answers is not in the document]
Comment (thomasyu888) [reply src/glue/jobs/s3_to_json_s3.py:726]: Ah right! I totally spaced about that...

**Pull request merged**